app/src/main/java/com/example/drawingactivity/DialogUtils.kt

```diff
@@ -0,0 +1,193 @@
+         */
+        onFailure: ((String) -> Unit)?,
+        /**
+         * This is Called when there is no Error and the Positive Button is pressed on the Dialog.
+         * This adds a Layer to a list.
+         */
+        onNameLayer: ((String) -> Unit)?,
+        /**
+         * Called when the Dialog Completes through any listener.
+         * This will reset all values to null in preparation for the next opening of the Dialog.
+         */
+        onDialogComplete: (() -> Unit)?,
+    ) {
+        val inflater = LayoutInflater.from(context)
+        val view = inflater.inflate(R.layout.add_layer_dialog, null)
+        val dialog = AlertDialog.Builder(context).apply {
+            setTitle(context.getString(R.string.addLayerDialogTitle))
+            setView(view)
+            val errorView = view.findViewById<TextView>(R.id.errorText)
+            val layerNameInput = view.findViewById<EditText>(R.id.layerNameField)
+            val currentHint: String = createLayerHint(context, layerList)
+            layerNameInput.hint = currentHint
+            var newLayerName = NAME_EMPTY
+            val currentNewLayerName: String? = onCheckForInput?.invoke()
+            if (currentNewLayerName != null) {
+                layerNameInput.setText(currentNewLayerName)
+                val errorStatus = checkForErrors(context, layerList, currentNewLayerName)
+                errorView.text = errorStatus.first
+                errorView.isVisible = errorStatus.second
+            }
+            layerNameInput.addTextChangedListener {
+                newLayerName = it.toString().trim()
+                onEditText?.invoke(newLayerName)
+                val errorStatus = checkForErrors(context, layerList, newLayerName)
+                errorView.text = errorStatus.first
+                errorView.isVisible = errorStatus.second
+            }
+            setPositiveButton(context.getString(R.string.addLayerPositiveButton)) { _, _ ->
+                if (errorView.isVisible) {
+                    onFailure?.invoke(errorView.text as String)
+                } else {
+                    when (newLayerName) {
+                        NAME_EMPTY, currentHint -> {
+                            newLayerName = currentHint
+                            onNameLayer?.invoke(newLayerName)
+                        }
+                        else -> {
+                            onNameLayer?.invoke(newLayerName)
+                        }
+                    }
+                }
+                onDialogComplete?.invoke()
+            }
+            setNegativeButton(context.getString(R.string.addLayerNegativeButton)) { _, _ ->
+                onDialogComplete?.invoke()
+            }
+            setOnCancelListener {
+                onDialogComplete?.invoke()
+            }
+        }
+        dialog.show()
+    }
+
+    fun layerListBottomSheet(
+        context: Context,
+        layerList: MutableList<String>,
```

Comment:
It's better to bring in just a `List` and then if you want to make a `Mutable` list locally you can